Base/Python/slicer/parameterNodeWrapper.py

```diff
@@ -0,0 +1,814 @@
+        parameterNode.UnsetParameter(name)
+
+    @staticmethod
+    def supportsCaching():
+        return True
+
+
+class NodeSerializer(Serializer):
+    """
+    Serializer for any instance (including subclasses) of slicer.vtkMRMLNode.
+    """
+    @staticmethod
+    def isIn(parameterNode, name: str) -> bool:
+        return parameterNode.GetNodeReferenceID(name) is not None
+
+    @staticmethod
+    def write(parameterNode, name: str, value: slicer.vtkMRMLNode) -> None:
+        return parameterNode.SetNodeReferenceID(name, value.GetID() if value is not None else None)
+
+    @staticmethod
+    def read(parameterNode, name) -> slicer.vtkMRMLNode:
+        return parameterNode.GetNodeReference(name)
+
+    @staticmethod
+    def remove(parameterNode, name) -> None:
+        """Removes this parameter from the node if it exists."""
+        parameterNode.RemoveNodeReferenceIDs(name)
+
+    @staticmethod
+    def supportsCaching():
+        return True
+
+
+class ObservedList(collections.abc.MutableSequence):
+    """
+    A list-like object that updates its associated parameter node when the container is updated.
+    Modification operations are supported (append, +=, __setitem__, etc), but non-modifying list
+    operations are not (+, *) as it would be too easy to accidentally make non-observable changes.
+    """
+    def __init__(self, parameterNode, listSerializer, name, startingValue):
+        self._parameterNode = parameterNode
+        self._serializer = listSerializer
+        self._name = name
+        self._list = startingValue
+
+    def __repr__(self) -> str:
+        return str(self)
+
+    def __str__(self):
+        return f"ObservedList({str(self._list)})"
+
+    def _saveList(self) -> None:
+        self._serializer.write(self._parameterNode, self._name, self._list)
+        # resetting the _list here helps if there are nested Multi's and someone does something like
+        #    m = parameterNode.multiMulti
+        #    m.append([1]) <-- this will seamlessly become an ObservedList so the next line works
+        #    m[0][0] = 2
+        self._list = self._serializer.read(self._parameterNode, self._name)._list
+
+    def __eq__(self, other) -> bool:
+        if isinstance(other, ObservedList):
+            return self._list == other._list
+        else:
+            return self._list == other
+
+    def __len__(self):
+        return self._list.__len__()
+
+    def __getitem__(self, index):
+        return self._list.__getitem__(index)
+
+    def __delitem__(self, index):
+        return self._list.__delitem__(index)
+
+    def __setitem__(self, index, item):
+        self._list.__setitem__(index, item)
+        self._saveList()
+
+    def __iadd__(self, other):
+        self._list.__iadd__(other)
+        self._saveList()
+        return self
+
+    def __add__(self, other):
+        raise NotImplementedError("Adding an ObservedList is not supported. However, += is supported.")
```

Comment:
Do these need to be forbidden? It seems clear to me that `data.the_list + data.the_other_list` should produce a _new_ list that is not observed.

I'm not sure how that would play with the caching mechanic, though.

I suppose `list(data.the_list) + ...` is a suitable workaround.

Reply:
In my opinion, the explicit breaking of the observation helps understandability. If a consensus of people think this should be implicit, I can change it.